pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -387,28 +387,23 @@ impl SerializedBatch {
     }
 
     pub(crate) fn from_values(batch: Vec<(CompactKey, Lsn, Value)>) -> Self {
-        use std::io::Write;
-
-        let mut offsets: Vec<(CompactKey, Lsn, u64)> = Vec::new();
-        let mut cursor = std::io::Cursor::new(Vec::<u8>::new());
+        // Pre-allocate a big flat buffer to write into. This should be large but not huge: it is soft-limited in practice by
+        // [`crate::pgdatadir_mapping::DatadirModification::MAX_PENDING_BYTES`]
+        let buffer_size = batch
+            .iter()
+            .map(|i| i.2.serialized_size().unwrap() as usize)
```

Comment:
We are serializing each value three times in this function. Feels a bit excessive.

Reply:
Yeah.  This is a bit transient because later we'll be building the buffer earlier, but we can clean it up here by calculating the serialized size much earlier and passing it around with the buffer: e94eb614ebcfcd3759d32391568714d83750eb70

(I checked that bloating those vectors with the serialized size doesn't degrade performance)